src/execution/stats.rs

```diff
@@ -15,63 +15,729 @@
-use log::info;
-use std::sync::Arc;
+use datafusion::{
+    datasource::physical_plan::ParquetExec,
+    physical_plan::{
+        aggregates::AggregateExec,
+        filter::FilterExec,
+        joins::{
+            CrossJoinExec, HashJoinExec, NestedLoopJoinExec, SortMergeJoinExec,
+            SymmetricHashJoinExec,
+        },
+        metrics::MetricValue,
+        projection::ProjectionExec,
+        sorts::{sort::SortExec, sort_preserving_merge::SortPreservingMergeExec},
+        visit_execution_plan, ExecutionPlan, ExecutionPlanVisitor,
+    },
+};
+use itertools::Itertools;
+use std::{sync::Arc, time::Duration};
 
 #[derive(Clone, Debug)]
 pub struct ExecutionStats {
-    bytes_scanned: usize,
-    // exec_metrics: Vec<ExecMetrics>,
+    query: String,
+    rows: usize,
+    batches: i32,
+    bytes: usize,
+    durations: ExecutionDurationStats,
+    io: Option<ExecutionIOStats>,
+    compute: Option<ExecutionComputeStats>,
+    plan: Arc<dyn ExecutionPlan>,
 }
 
 impl ExecutionStats {
-    pub fn bytes_scanned(&self) -> usize {
-        self.bytes_scanned
+    pub fn try_new(
+        query: String,
+        durations: ExecutionDurationStats,
+        rows: usize,
+        batches: i32,
+        bytes: usize,
+        plan: Arc<dyn ExecutionPlan>,
+    ) -> color_eyre::Result<Self> {
+        Ok(Self {
+            query,
+            durations,
+            rows,
+            batches,
+            bytes,
+            plan,
+            io: None,
+            compute: None,
+        })
+    }
+
+    pub fn collect_stats(&mut self) {
+        if let Some(io) = collect_plan_io_stats(Arc::clone(&self.plan)) {
+            self.io = Some(io)
+        }
+        if let Some(compute) = collect_plan_compute_stats(Arc::clone(&self.plan)) {
+            self.compute = Some(compute)
+        }
+    }
+
+    pub fn rows_selectivity(&self) -> f64 {
+        let maybe_io_output_rows = self.io.as_ref().and_then(|io| io.parquet_output_rows);
+        if let Some(io_output_rows) = maybe_io_output_rows {
+            self.rows as f64 / io_output_rows as f64
+        } else {
+            0.0
+        }
+    }
+
+    pub fn bytes_selectivity(&self) -> f64 {
+        let maybe_io_output_bytes = self.io.as_ref().and_then(|io| io.bytes_scanned.clone());
+        if let Some(io_output_bytes) = maybe_io_output_bytes {
+            self.bytes as f64 / io_output_bytes.as_usize() as f64
+        } else {
+            0.0
+        }
+    }
+
+    /// A ratio of the selectivity of the query to the effectivness of pruning the parquet file.
+    ///
+    /// V1: Simply look at row groups pruned by statistics and the row selectivity
+    /// TODO: Incorporate bloom filter and page index pruning
+    ///
+    /// Example calculations:
+    ///
+    /// 1. No pruning and select all rows
+    ///    - row groups pruned: 0
+    ///    - row groups matched: 10
+    ///    - row selectivity: 1.0
```

Comment:
I am mocking out some calculations here to figure out how i want to compute "efficiency" - i just have a general idea that i want to take into account row selectivity and parquet pruning but need to flesh out the details and some tangible use cases.

Reply:
Ended up making a [spreadsheet](https://docs.google.com/spreadsheets/d/1wDqe-oGHg9SDL_7VoIinvYoYV1Svst_5G5YTS2IFQ8g/edit?gid=0#gid=0) for this instead.